Fix PiCT reported bugs + build enhancements (13).

swoocn opened the pull request:
- Landing page: mod Seller marker popup size consistency and positioning. 
- User Preferences navbar: make language selection section scrollable.
- Reviews + Reply to Review screen: polish up save button/ message behavior.
- Removed flag icons from Language List feature.
- Added Sanctioned Sellers reporting. 
- Introduce Traditional Chinese language.
- Introduce Vietnamese language.
- Update Benin Ewe language.
- Update App Version to 'v1.4.0'.

**[see commits]**

_This PR is subject to change if additional fixes + enhancements are introduced before approval._

swoocn commented:
Hi @tate1650 - Could you please keep this PR open and refrain from merging until I give the word?
Thanks!

swoocn commented:
Hi @tate1650 - 
Hope you are doing well!
I wanted to inform you that the new **Sanctioned Sellers Report** can now be triggered via the `/api/v1/reports/sanctioned-sellers-report` endpoint.  To make this process straightforward, you can configure and execute the request directly using the Swagger UI in the staging/ prod environments.  Below are some screenshots showing an example response in our Vercel BE environment ->

![image](https://github.com/user-attachments/assets/f25a249c-ba1c-44da-8386-af3205656e71)
![image](https://github.com/user-attachments/assets/acd6b1c7-df66-41e0-aa88-4e78c11914eb)

Before running the report, please note that you will need to add a new **sanctioned-region** collection in MongoDB which will contain the polygon boundary coordinates for each sanctioned region.  I've attached a script to help build and populate the collection.

Please let me know how the setup and execution process goes.  If you have any questions or encounter any issues, feel free to reach out!

Thanks again!

tate1650 commented:
Hi Danny, thanks for the info on the sanctioned seller report feature. Just to double-check, is this PR ready for a review or should I still wait for a more explicit thumbs-up?

swoocn commented:
> Hi Danny, thanks for the info on the sanctioned seller report feature. Just to double-check, is this PR ready for a review or should I still wait for a more explicit thumbs-up?

I will let you know shortly @tate1650.  Thx!

swoocn commented:
> > Hi Danny, thanks for the info on the sanctioned seller report feature. Just to double-check, is this PR ready for a review or should I still wait for a more explicit thumbs-up?
> 
> I will let you know shortly @tate1650. Thx!

Hi @tate1650 - Please review and approve the changes when you can so we can proceed with the deployment propagation. Thx!

tate1650 reviewed (changes requested):
This looks pretty good, but for the sanctioned country detection, I was anticipating something more along the lines of an approach to proactively _prevent_ users from creating shops in sanctioned countries. I.e. have some sort of check in place during the sell-center-setting stage to stop the user from setting the sell center in a place they shouldn't be able to.

I recognize that use of the Nominatim API is limited, so an approximate approach to this would be fine (e.g. the bounding polygon approach with more precise shapes). Then we could use the `getSanctionedSellersReport` controller action for cleanup of any stragglers. The idea here being that we're not making the report/cleanup our first and only line of defense. Thoughts?

swoocn commented:
> This looks pretty good, but for the sanctioned country detection, I was anticipating something more along the lines of an approach to proactively _prevent_ users from creating shops in sanctioned countries. I.e. have some sort of check in place during the sell-center-setting stage to stop the user from setting the sell center in a place they shouldn't be able to.
> 
> I recognize that use of the Nominatim API is limited, so an approximate approach to this would be fine (e.g. the bounding polygon approach with more precise shapes). Then we could use the `getSanctionedSellersReport` controller action for cleanup of any stragglers. The idea here being that we're not making the report/cleanup our first and only line of defense. Thoughts?

Thank you for the feedback, @tate1650! 
Sorry, I should have clarified that this entire sanctioned sellers initiative is being approached in two phases.
Phase 1 focuses on reporting to identify sellers currently in sanctioned regions and to assess the reliability/ integrity of the Nominatim API.  Once we’re confident in the accuracy of the data after running the report in prod, we’ll move onto Phase 2.  This phase will involve proactively preventing sellers from selling in sanctioned regions by hiding their markers.  The report will then  serve as a supplementary tool to periodically check for any stragglers.  
Feel free to share your thoughts!

swoocn commented on backend/src/controllers/reportsController.ts:
```diff
@@ -0,0 +1,17 @@
+import {reportSanctionedSellers} from "../services/report.service";
+import logger from "../config/loggingConfig";
+
+export const getSanctionedSellersReport = async (req: Request, res: Response) => {
```
> This strikes me as the sort of thing that we would want to make a privileged resource, as opposed to letting it be accessible for any user to use.
> 
> Can we perhaps either:
> 
> * Set this up in such a way that it can be run as a node script from the BE server, OR
> * Add some kind of user permission flag that regulates whether someone can use this controller action?
> 
> I'm fine with either (or even a third alternative option if you can think of one); just don't want this to be freely available to anyone.

Yes, I agree that the endpoint should not be freely accessible.  I can certainly add some basic auth middleware to better secure and restrict access.

swoocn commented on backend/src/models/enums/restrictedArea.ts:
```diff
@@ -0,0 +1,83 @@
+  LUHANSK_OBLAST = "Luhansk Oblast"
+}
+
+export const RestrictedAreaBoundaries = {
+  [RestrictedArea.CUBA]: {
+    type: "Polygon",
+    // (longitude, latitude)
+    coordinates: [[
+      [-85.3, 19.4], // bottom-left corner 
+      [-73.8, 19.4], // bottom-right corner
+      [-73.8, 23.7], // top-right corner
+      [-85.3, 23.7], // top-left corner
+      [-85.3, 19.4], // close the polygon shape
+    ]],
+  },
```
That is correct!
To ensure a performant implementation, we will first filter sellers located in or near the bounding box of the sanctioned region and then use the Nominatim API for validation.  Feel free to share any thoughts for improvement.

tate1650 replied:
That approach seems pretty reasonable to me. Just wanted to make sure I was understanding for review's/curiosity's sake. :)

swoocn commented on backend/src/routes/report.routes.ts:
```diff
@@ -0,0 +1,30 @@
+ *       500:
+ *         description: Internal server error
+ */
+reportRoutes.post(
```
> I'm thinking this should probably be a GET request, since from what I saw it doesn't really perform any write operations.

I can see why you think that.
We actually _write_ the report to the Sentry.io platform in the form of `logger.error`, in the shared environments.  I'm indifferent either way, between GET vs POST.

tate1650 replied:
Ah okay gotcha. I think what's happening to the _app state_ (i.e. being read or written) is probably more immediately relevant when it comes to which HTTP method to use, so I'd recommend GET. But it's also not that important (more of a style/convention thing) so feel free to either change or ignore. Definitely more of a nitpick.

tate1650 commented on frontend/src/constants/menu.ts:
```diff
@@ -55,30 +55,38 @@ export const menu = {
+        label: 'ZH/TW',
+        title: 'Traditional Chinese',
+        translation: '繁體中文',
+        icon: '/images/shared/language/zh.svg',
```
Ah I see. I didn't realize that it was the same image file for both versions of Chinese.

So I'm by no means an expert on written Chinese linguistics, but from what I've read it seems like Traditional Chinese isn't really used much in the PRC, so it might make sense to have a different flag icon for a region where it's more commonly used (e.g. Taiwan or Hong Kong). Thoughts?

tate1650 commented:
Ah okay, that sounds good to me! Thank you for clarifying the plan. So in that case, for this PR, the only changes I'd request then are to add some kind of auth protection for the report, and possibly the flag image depending on your thoughts.

Let me know when we're ready for another round of review!

swoocn commented:
> Ah okay, that sounds good to me! Thank you for clarifying the plan. So in that case, for this PR, the only changes I'd request then are to add some kind of auth protection for the report, and possibly the flag image depending on your thoughts.
> 
> Let me know when we're ready for another round of review!

Hi @tate1650 - 
Please review the latest changes as follows ->  
- **Language List**; I went ahead and removed the flag icons from the language selection list to prevent any potential conflicts/ controversies in the future.
- **Report API endpoint**; I've added a basic authentication middleware to the report routing.  As part of this, two new environment variables - `ADMIN_API_USERNAME` and `ADMIN_API_PASSWORD` - have been introduced.
I can email you the credentials we've already set up in our Netlify environment; alternatively, you're welcome to create your own credentials,.  

Please feel free to reach out if you have any questions.  Thanks again!

tate1650 reviewed (changes requested):
Hey Danny!

> * **Language List**; I went ahead and removed the flag icons from the language selection list to prevent any potential conflicts/ controversies in the future.

Sure, sounds good to me!

> * **Report API endpoint**; I've added a basic authentication middleware to the report routing.  As part of this, two new environment variables - `ADMIN_API_USERNAME` and `ADMIN_API_PASSWORD` - have been introduced. I can email you the credentials we've already set up in our Netlify environment; alternatively, you're welcome to create your own credentials,.

So if it's okay for only CT to run this endpoint, this approach should be fine. However, if your team is trying to use this endpoint, we're going to hit some logistical issues because CT will generate credentials separately for each environment and we will not be able to share them with you directly. If that's okay, then we can proceed with this approach. Otherwise, we might need to look into an alternate solution that would allow your team members to access the endpoint with auth protections, but without needing access to a secret environment variable.

Lemme know what you think about all this, and please don't hesitate to ask follow-up questions if you have any!

swoocn commented on backend/src/middlewares/verifyToken.ts:
```diff
@@ -48,4 +48,30 @@ export const verifyToken = async (
   }
+};
+
+export const verifyAdminToken = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
```
Thanks again for the feedback, @tate1650!
We’re comfortable moving forward with this approach and were planning to rely on you/ PiCT to verify the integrity of  our implemented process - as we have tested it ourselves in our own Test environment. Ultimately, the goal is to establish Phase 2 and use the reporting as a supplementary tool executed on a more or less periodic basis. If you're okay with it, we can coordinate with you to run report as needed. Perhaps down the line, we can look into adjusting the process so we can manage it ourselves..

Let me know your thoughts, and thanks again for your support!

tate1650 replied:
Alright, yep, this sounds good to me!

tate1650 commented on backend/src/middlewares/verifyToken.ts:
```diff
@@ -48,4 +48,30 @@ export const verifyToken = async (
+  const base64Credentials = authHeader && authHeader.split(" ")[1];
+  if (!base64Credentials) {
+    logger.warn("Admin credentials are missing.");
+    return res.status(401).json({ message: "Admin credentials are required" });
```
We should be a little more vague/generic with the error messages for the 401 responses (e.g. "Authentication error")

swoocn replied:
Good catch!  I will get it resolved.

tate1650 reviewed (approved):
Gonna go ahead and mark this as approved, but if we can fix up those 401 messages real quick before merging that would be awesome! Missed that in my previous review and randomly realized it during the day today 😆.

swoocn commented:
Hi @tate1650 - Please merge when you can so we can proceed with the deployment propagation. Thx!

tate1650 commented:
Alright, it's a bit late in the work day here so I'll deploy this to staging to start with, so we can validate and then propagate to prod tomorrow.

tate1650 commented:
Changes have been deployed to staging. I can help validate the ADMIN_API stuff tomorrow, but if you could validate all the UI-accessible stuff and let me know how it goes that would be awesome!

swoocn commented:
> Changes have been deployed to staging. I can help validate the ADMIN_API stuff tomorrow, but if you could validate all the UI-accessible stuff and let me know how it goes that would be awesome!

Yup, I will keep you posted; thanks!

tate1650 commented:
> Before running the report, please note that you will need to add a new sanctioned-region collection in MongoDB which will contain the polygon boundary coordinates for each sanctioned region. I've attached a script to help build and populate the collection.

Hey, trying to run this on staging (apologies for the delay), but I'm not seeing a dedicated script. Did you mean to send it to me directly? Or when you mentioned the script were you referring to the `jest.setup.ts` file?

swoocn commented:
Hi @tate1650 - No worries; I sent you the script via email just now.  Thanks again! 😎👍
